pico/pico_cmn.c

```diff
@@ -150,23 +149,6 @@ static int PicoFrameHints(void)
       }
     }
 
-    // get samples from sound chips
-    if ((y == 224 || y == line_sample) && PicoIn.sndOut)
-    {
-      cycles = SekCyclesDone();
-
-      if (Pico.m.z80Run && !Pico.m.z80_reset && (PicoIn.opt&POPT_EN_Z80))
-        PicoSyncZ80(cycles);
-#ifdef PICO_CD
-      if (PicoIn.AHW & PAHW_MCD)
-        pcd_sync_s68k(cycles, 0);
-#endif
-#ifdef PICO_32X
-      p32x_sync_sh2s(cycles);
-#endif
-      PsndGetSamples(y);
-    }
-
     // Run scanline:
```

Comment:
What allows you to do that?

The thing is sampling the chips once per frame is insufficient to have correct sounding audio. This hack of doing it around the vsync time made it sort of work because the games usually do the sound chip programming in vsync interrupt handlers.

Reply:
> What allows you to do that?
> 
> The thing is sampling the chips once per frame is insufficient to have correct sounding audio. This hack of doing it around the vsync time made it sort of work because the games usually do the sound chip programming in vsync interrupt handlers.

AFAIK the old mechanism only related to FM sound - PSG and DAC where done on the fly on register change, and X32 and CD sounds where only added at the rendering points.
There where issues with this in some roms, e.g. Daimakaimura having a noticable "ping" sound at the end of the title screen. That was due to the FM calculated only after registers had been written twice over. The idea of my mechanism is to calculate FM sound when registers are written to, by means of calling PsndDoFM in pico/memory.c. Since the FM registers are often changed in a block, a hysteresis has been added - some 4ms. That solved those sound issues for me. It's unfinished though, I'm not yet totally satisfied with it. It produces other effects I would rather like to avoid, and it's still called too often for my taste.

Have I overlooked something here? Is this idea somehow bad?

Reply:
Sounds good, thanks.